pyop2/base.py

```diff
@@ -1363,154 +1283,59 @@ def __repr__(self):
-        dat._recv_buf.clear()
+    @abc.abstractproperty
+    def local_to_global_numbering(self):
+        """The mapping from process-local to process-global numbers for this halo."""
+        pass
 
-    @property
-    def sends(self):
-        """Return the sends associated with this :class:`Halo`.
+    @abc.abstractmethod
+    def global_to_local_begin(self, dat, insert_mode):
+        """Begin an exchange from global (assembled) to local (ghosted) representation.
 
-        A dict of numpy arrays, keyed by the rank to send to, with
-        each array indicating the :class:`Set` elements to send.
+        :arg dat: The :class:`Dat` to exchange.
+        :arg insert_mode: The insertion mode.
+        """
+        pass
 
-        For example, to send no elements to rank 0, elements 1 and 2 to rank 1
-        and no elements to rank 2 (with ``comm.size == 3``) we would have: ::
+    @abc.abstractmethod
+    def global_to_local_end(self, dat, insert_mode):
+        """Finish an exchange from global (assembled) to local (ghosted) representation.
 
-            {1: np.array([1,2], dtype=np.int32)}.
+        :arg dat: The :class:`Dat` to exchange.
+        :arg insert_mode: The insertion mode.
         """
-        return self._sends
-
-    @property
-    def receives(self):
-        """Return the receives associated with this :class:`Halo`.
+        pass
 
-        A dict of numpy arrays, keyed by the rank to receive from,
-        with each array indicating the :class:`Set` elements to
-        receive.
+    @abc.abstractmethod
+    def local_to_global_begin(self, dat, insert_mode):
+        """Begin an exchange from local (ghosted) to global (assembled) representation.
 
-        See :func:`Halo.sends` for an example.
+        :arg dat: The :class:`Dat` to exchange.
+        :arg insert_mode: The insertion mode.
         """
-        return self._receives
-
-    @property
-    def global_to_petsc_numbering(self):
-        """The mapping from global (per-process) dof numbering to
-    petsc (cross-process) dof numbering."""
-        return self._global_to_petsc_numbering
+        pass
 
-    def verify(self, s):
-        """Verify that this :class:`Halo` is valid for a given
-:class:`Set`."""
-        for dest, sends in self.sends.items():
-            assert (sends >= 0).all() and (sends < s.size).all(), \
-                "Halo send to %d is invalid (outside owned elements)" % dest
+    @abc.abstractmethod
+    def local_to_global_end(self, dat, insert_mode):
+        """Finish an exchange from local (ghosted) to global (assembled) representation.
 
-        for source, receives in self.receives.items():
-            assert (receives >= s.size).all() and \
-                (receives < s.total_size).all(), \
-                "Halo receive from %d is invalid (not in halo elements)" % \
-                source
+        :arg dat: The :class:`Dat` to exchange.
+        :arg insert_mode: The insertion mode.
+        """
+        pass
```

Comment:
We basically only ever use the Firedrake halo, so make this an abstract class and change the naming convention to more closely match PETSc.